pykoopman/regression/_edmdc.py

```diff
@@ -0,0 +1,138 @@
+    that satisfy y' = Ay + Bu and x = Cy, where y' is the time-shifted
+    observable with y0 = phi(x0) and u is the control input. B and C
+    are the unknown control and measurement matrices, respectively.
+
+    The objective functions,
+    :math:`\\|Y'-AY-BU\\|_F`
+    and
+    :math:`\\|X-CY\\|_F`,
+    are minimized using least-squares regression and singular value
+    decomposition/
+
+    See the following reference for more details:
+
+        `Korda, M. and Mezic, I.
+        "Linear predictors for nonlinear dynamical systems:
+        Koopman operator meets model predictive control."
+        Automatica, Vol. 93, 149–160.
+        <https://www.sciencedirect.com/science/article/abs/pii/S000510981830133X>`_
+
+    Parameters
+    ----------
+
+    Attributed
+    ----------
+    coef_ : array, shape (n_input_features_, n_input_features_) or
+        (n_input_features_, n_input_features_ + n_control_features_)
+        Weight vectors of the regression problem. Corresponds to either [A] or [A,B]
+
+    state_matrix_ : array, shape (n_input_features_, n_input_features_)
+        Identified state transition matrix A of the underlying system.
+
+    control_matrix_ : array, shape (n_input_features_, n_control_features_)
+        Identified control matrix B of the underlying system.
+
+    projection_matrix_ : array, shape (n_input_features_+n_control_features_, svd_rank)
+        Projection matrix into low-dimensional subspace.
+
+    projection_matrix_output_ : array, shape (n_input_features_+n_control_features_,
+                                              svd_output_rank)
+        Projection matrix into low-dimensional subspace.
+
+     Examples
+    --------
+    TODO
+    """
+
+    def __init__(self):
+        pass
+
+    def fit(self, x, y=None, u=None, dt=None):
+        """
+        Parameters
+        ----------
+        x: numpy ndarray, shape (n_samples, n_features)
+            Measurement data to be fit.
+
+        u: numpy.ndarray, shape (n_samples, n_control_features), \
+                optional (default None)
+            Time series of external actuation/control.
+
+        y: numpy.ndarray, shape (n_samples, n_features)
+            Time-shifted measurement data to be fit
+
+        dt: scalar
+            Discrete time-step
+
+        Returns
+        -------
+        self: returns a fitted ``EDMDc`` instance
+        """
+        self.n_samples_, self.n_input_features_ = x.shape
+        if y is None:
+            X1 = x[:-1, :]
+            X2 = x[1:, :]
+        else:
+            X1 = x
+            X2 = y
+
+        if u.ndim == 1:
+            if len(u) > X1.shape[0]:
+                u = u[:-1]
+            C = u[np.newaxis, :]
+        elif u.ndim == 2:
+            if u.shape[0] > X1.shape[0]:
+                u = u[:-1, :]
+            C = u
+        self.n_control_features_ = C.shape[1]
+
+        self._fit(X1, X2, C)
+        return self
+
+    def _fit(self, X1, X2, U):
```

Comment:
again, here the .fit is not using any fitting from PyDMD. I thought we are actually appreciating those fancy fitting procedures in PyDMD, say some noise reduction SVD, some Tihokniv regularization. 

I would of course like to implement the pykoopman in the simple way, i.e., just call LS from numpy. But from our previous discussion, there are some opinions insisting on using pydmd for regression. What shall we do now? 

One dirty solution is to copy-paste-modify the regression steps in pydmd.
+: we can go beyond pydmd's restriction on using consecutive data.
+: we will not be influenced if pydmd has some version changes.
-: we cannot say we build our package that connects to the functions in pydmd, which is a nice way to acknowledge other's work.

Reply:
This depends on how you would like to develop the package. As you pointed out there are some advantages + disadvantages replacing the use of pydmd with your own code. I would add to your list:
- requires more time due to implementing new functionality
- requires more time for maintenance

Your first (and to some extent second) is the major point here, since it is quite restrictive and prohibits the use of methods, which make use of non-consecutive data. To circumvent the issue, Brian and I decided on a structure that allows both: the user can specify a pydmd regressor that will be used within pykoopman, but we also implement regressors directly into pykoopman, e.g. edmdc, that are not available within pydmd.

It's not really necessary to replace pydmd, just add your own regressors in the regressors folder. pykoopman in its current form works with both. But if you prefer either way, you can absolutely go for it. I am nearly finished with my code contributions and then feel free to make these big refactor changes.